Installing-and-Configuring/OS/index.md

```diff
@@ -0,0 +1,136 @@
+
+To mitigate recovery shell attacks, entering the recovery shell will modify the PCR and measurements. Also changes to kernel or initrd files on /boot will modify the measurements.  This will be detected as tampering.
+
+See also [TPM vs security dongle](#tpm-vs-security-dongle)
+
+#### Disable TPM Disk Unlock passphrase/key
+
+This feature may be turned off during configuration of heads.  The following will instruct heads not to ask for a disk unlock passphrase and to not store a disk unlock key in the TPM.  
+
+  CONFIG_TPM_NO_LUKS_DISK_UNLOCK=y
+
+### Using Heads Disk Recovery Key 
+
+In the Disk Recovery Key scenario, Heads doesn't attempt to release the Disk Unlock Key. It simply relies on OS mechanisms to directly prompt the user for the Disk Recovery Key passphrase at standard OS prompt.  
+
+This model is vulnerable to brute force attacks.
+
+### Using an OS managed LUKS passphrase (not heads)
+
+In this model heads is not involved in processes related to the decryption of data on disk. The OS manages keys and access to the LUKS containers by asking the user for a passphrase on boot.  
+
+This model is vulnerable to brute force attacks.
+
+### Security Dongle with LUKS root (not heads)
+
+In this model heads is not responsible for any of the decryption of data on disk.  You may use a security dongle (aka smartcard aka security token) to manage the passphrase for the root LUKS partition.  Like the TPM, the security dongle enforces limits which prevent brute force attacks.
+
+In contrast to [using TPM for decryption](#using-tpm-for-decryption-of-drives)--where a disk unlock passphrase is stored in the TPM--in this model the TPM is not used for LUKS decryption.  A LUKS disk unlock key/passphrase is encrypted and stored in a 'public' location such as /boot.  After heads boots the kernel a process which uses the security dongle decrypts the LUKS root partition.
+
+See also [TPM vs security dongle](#tpm-vs-security-dongle)
+
+TPM vs Security Dongle
+----
+
+Which one you choose depends on hardware you have and threat models you are working with.  Your system may not have a TPM or you may not have a smartcard.  The core functionality is similar in that they both store private keys and allow for cryptographic operations.
+
+Heads uses the TPM for various operations and may use the TPM for disk unlock.  The significant differences between a TPM based Disk Unlock Key and USB Security dongle are
+
+* they are two different proprietary hardware chips
+* TPM has access to RAM and processor so it can 'measure' system states
+* Security Dongles often have other features such as password management
+* the TPM is a non-removable chip inside the system but the USB Security dongle is removable
```

Comment:
"TPM is a non-removable chip" that is not true for KGPE-D16 and other workstations/servers platforms. You can say that it is a chip, soldered in most cases, to the motherboard. A picture of a removable TPM is showed https://osresearch.net/Keys/#tpm-pcrs

"TPM has access to RAM and processor". Hmm. 

The TPM, is initialized as early as possible by coreboot, normally in bootblock or romstage, and is fed content to be measured through a tool/library. Heads then use userland tool to ask for measurements that already happened earlier, where it can extends those measurements. Loading kernel modules under Heads calls a wrapper around insmod which extends PCR#5. Going into recovery extends PCR#4. And having inconsistent measurements won't permit to unseal TPM NV's own memory.


So to reflect reality here, while it is true that the TPM is not completely independent of the system (it shares LPC bus in most cases) it has limited access to external resources, you can communicate with the TPM. The TPM can read what is on that LPC (or other bus it is connected to) but stating that:

"TPM has access to RAM and processor so it can 'measure' system states" leaves the user thinking that the TPM is having the capacity to read states, which is misleading. The TPM measures content passed to it. Stores values, extends values. Can generate keys, secrets, can be used as a random numbr generator etc but all of this is out of scope here. Once again, explaining so that we are on the same page technically.

 So 
 
 TPM is fed content to measure and normally store the result of those measurements in what is called [PCRs](https://osresearch.net/Keys/#tpm-pcrs).

If you want to go into that:
In the case of Intel TXT (not covered by Heads right now), Intel's ACM binary blob is fed to the processor to be executed first, which is used as a Hardware Root of Trust (RoT). Since the ACM blob is verified against fused Intel signature, [the ACM is trusted into reporting the first measurement into PCR0 per coreboot's IBB/CRTM measured boot implementation](https://doc.coreboot.org/security/vboot/measured_boot.html#ibb-crtm).  Heads uses coreboot measured boot in [SRTM mode](https://doc.coreboot.org/security/vboot/measured_boot.html#srtm-mode) where the schematic can be observed [here](https://doc.coreboot.org/security/vboot/measured_boot.html#srtm-mode).

As always, I would keep things simple and refer externally to have better explanations instead of bloating things here. But what I want you to understand here is that Heads aligned measurements scheme with Coreboot, and that all coreboot related measurements are happening in [PCR#2 ](https://doc.coreboot.org/security/vboot/measured_boot.html#pcr-2) per this [commit](https://github.com/osresearch/heads/commit/28d3b7c89c114fccdf937f55ccce4397fa9338d1) on boot (coreboot 4.8.1 and forward) while [Heads feeds the TPM to measure other Heads stuff](https://github.com/osresearch/heads/commit/28d3b7c89c114fccdf937f55ccce4397fa9338d1) for example [here](https://github.com/osresearch/heads/search?q=%22tpm+extend%22&type=code) and [here](https://github.com/osresearch/heads/blob/c3b0bd6ffbe816430dd41ef54e649af52ed1ff3b/initrd/bin/kexec-seal-key#L85) to measure user related configs and public key, LUKS header, kernel modules etc.